lib/constructBodyDigestSignature.js

```diff
@@ -0,0 +1,60 @@
+const crypto = require('crypto')
+
+const constructBodyDigestSignature = ({
+  headers: {
+    digest,
+    'content-type': contentType,
+    'content-encoding': contentEncoding
+  },
+  body
+}) => {
+  const { formattedAlgorithm, digestAlgorithm, digestValue } = parseBodyDigestHeader(digest)
+
+  const calculatedDigest = calculateDigest({
+    algorithm: formattedAlgorithm,
+    message: parseContentType(contentType, contentEncoding)(body),
+    encoding: lookupEncoding(formattedAlgorithm)
+  })
+
+  if (calculatedDigest !== digestValue) {
+    throw new Error('Calculated Digest does not match Digest from Request Header')
+  }
+
+  return `${digestAlgorithm}=${calculatedDigest}`
+}
+
+const parseBodyDigestHeader = (requestDigestHeader) => {
+  // TODO:(jeff-sexton) - Parse for multiple digests in comma separated list and select one
+  const [digestAlgorithm, digestValue] = requestDigestHeader.split(/=(.+)/)
+  const formattedAlgorithm = digestAlgorithm.toLowerCase().split('-').join('')
+
+  return { formattedAlgorithm, digestAlgorithm, digestValue }
+}
+
+const parseContentType = (requestContentTypeHeader, requestContentEncodingHeader) => {
+  // TODO:(jeff-sexton) - Add other media types and support for ContentEncoding
+  const mediaTypeFuncMap = {
+    'application/json': (resource) => JSON.stringify(resource),
+    default: () => { throw new Error('Unrecognized media-type') }
+  }
+
+  if (!requestContentTypeHeader) {
+    throw new Error('No Content-Type Header in request')
+  }
```

Comment:
Should this validation be handled outside of the digest construction? We probably want this method to only handle the digest calculations. 

There are other validations we may want to add, so lets think about a central place we can add them and allow them to be overridden.

Reply:
This validation is no longer necessary in the streamlined `getDigest`